source/session_encrypt.c

```diff
@@ -183,8 +191,14 @@ static int sign_header(struct aws_cryptosdk_session *session) {
     // see what happened. It also makes sure that the header is fully initialized,
     // again just in case some bug doesn't overwrite them properly.
 
-    memset(session->header.iv.buffer, 0x42, session->header.iv.len);
-    memset(session->header.auth_tag.buffer, 0xDE, session->header.auth_tag.len);
+    if (session->header.iv.len) {
+        assert(session->header.iv.buffer);
+        memset(session->header.iv.buffer, 0x42, session->header.iv.len);
+    }
+    if (session->header.auth_tag.len) {
+        assert(session->header.auth_tag.buffer);
+        memset(session->header.auth_tag.buffer, 0xDE, session->header.auth_tag.len);
+    }
```

Comment:
Should we assert or return error? @alex-chew

Reply:
Short answer: assert. The condition should always be true, and if not then it's a programming error rather than a user/data error.

Long answer: I think we can actually leave out the assert, since the `aws_cryptosdk_hdr_is_valid(&session->header)` precondition implies that the `session->header.iv` and `session->header.auth_tag` are `aws_byte_buf_is_valid`, which in turn implies that `iv.len` implies `iv.buffer` and likewise for `auth_tag`. But the assert is valuable as documentation of this intent.

Reply:
Digging deeper, it looks like the only callsite of `sign_header` appears after a successfull call to `build_header`, which initializes these buffers to have positive length (since all supported algorithm suites use 12- and 16-byte IVs and authentication tag lengths, respectively). So we can change the guards/asserts to preconditions (i.e. `byte_buf_is_valid` and `buf.len > 0` for each of the IV and auth tag).